src/validate.rs

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) 2022 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use anyhow::{anyhow, Result};
+use oci_distribution::Reference;
+use serde::{Deserialize, Serialize};
+use signature::{Image, Policy, SignatureScheme};
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+
+use get_resource::get_resource_service_client::GetResourceServiceClient;
+use get_resource::GetResourceRequest;
+
+const AA_GETRESOURCE_ADDR: &str = "http://127.0.0.1:50001";
+
+// Image security config dir contains important information such as
+// security policy configuration file and signature verification configuration file.
+// Therefore, it is necessary to ensure that the directory is stored in a safe place.
+//
+// The reason for using the `/run` directory here is that in general HW-TEE,
+// the `/run` directory is mounted in `tmpfs`, which is located in the encrypted memory protected by HW-TEE.
+const IMAGE_SECURITY_CONFIG_DIR: &str = "/run/image-security";
+const POLICY_FILE_PATH: &str = "/run/image-security/security_policy.json";
+
+mod get_resource {
+    tonic::include_proto!("getresource");
+}
+
+#[derive(EnumString, Display, Debug, PartialEq)]
+pub enum SimpleSigning {
+    #[strum(to_string = "/run/image-security/simple_signing")]
+    ConfigDir,
+    #[strum(to_string = "/run/image-security/simple_signing/sigstore_config")]
+    SigstoreConfigDir,
+    #[strum(to_string = "/run/image-security/simple_signing/keyring.gpg")]
+    GpgKeyRing,
+}
+
+/// The resource description that will be passed to AA when get resource.
+#[derive(Serialize, Deserialize, Debug)]
+struct ResourceDescription {
+    name: String,
+    optional: HashMap<String, String>,
+}
+
+impl ResourceDescription {
+    /// Create a new ResourceDescription with resource name.
+    pub fn new(name: &str) -> Self {
+        ResourceDescription {
+            name: name.to_string(),
+            optional: HashMap::new(),
+        }
+    }
+}
+
+/// `security_validate` judges whether the container image is allowed to be pulled and run.
+///
+/// According to the configuration of the policy file, if the signature
+/// of the container image needs to be verified, the specified signature
+/// scheme is used for signature verification.
+pub async fn security_validate(
+    image_reference: &str,
+    image_digest: &str,
+    aa_kbc_params: &str,
+) -> Result<()> {
+    if !Path::new(IMAGE_SECURITY_CONFIG_DIR).exists() {
+        fs::create_dir_all(IMAGE_SECURITY_CONFIG_DIR)
+            .map_err(|e| anyhow!("Create image security runtime config dir failed: {:?}", e))?;
+    }
+
+    // if Policy config file does not exist, get if from KBS.
+    if !Path::new(POLICY_FILE_PATH).exists() {
```

Comment:
Whether need to add a mutex here to ensure only one thread will invoke this `get_resource_from_kbs()` ?
Like this scenario:
Kata-agent receives two image pulling requests (Request_1 and Request_2) simultaneous after the new pod is created.
1. image-rs handle the Request_1 firstly and the POLICY_FILE_PATH is empty this time.
2. So it will enter the if() and execute `get_resource_from_kbs("Policy", aa_kbc_params).await?` 
3. Because the `get_resource_from_kbs()` need to communicate to kbs and will cost much time, and the `.await?` is also means a sync thread switch point. So this thread maybe switched here.
4. Then the image-rs handle the Request_2 request, and it will also find the POLICY_FILE_PATH is empty.
5. So the Request_2 corresponding thread will also execute the same `get_resource_from_kbs()`.